### PR TITLE
feat(engine): add retry backoff support in BPMN

### LINKED ISSUE #11101: I as a Zeebe user want to set up a backoff timeout for failed tasks in a BPMN XML.
**Is your feature request related to a problem? Please describe.**
In our cases, it would be cool to set up a constant backoff timeout if a task failed (i.e. between retries) in the BPMN XML.

**Describe the solution you'd like**
I want to be able to set a backoff interval in the XML, somewhere near the retries.

**Describe alternatives you've considered**
We could add backoff support in the retries field. I.e. R3PT7m - retry 3 times with a delay of 7 minutes.

### PR DESCRIPTION
**Pull request by aivinog1**
## Description

I've added retry backoff support in BPMNs for service tasks.

## Related issues

closes #11101 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [X] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [X] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Operate](https://github.com/camunda/operate/issues)
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

**Comment by Zelldon** [comment]
Hey @aivinog1 thanks for your contribution, please be aware that most of the team is on holiday so it might take a while until we can review your PR (likely until the end of next week). Sorry for the inconvenience:bow:

**Comment by aivinog1** [comment]
Hi @Zelldon! 
Can we speed up the review a little bit in terms of BPMN API (I mean the XML structure)? We could settle down the API and I could create a PR in the Modeler. Thanks 🙂

**Comment by Zelldon** [comment]
This week @remcowesterhoud and @saig0 are back

**Comment by korthout** [comment]
Sorry for the long wait @aivinog1 I'll have a look at it now.

File: bpmn-model/src/main/java/io/camunda/zeebe/model/bpmn/util/time/Interval.java [hunk 1]
```diff
@@ -175,6 +176,28 @@ private boolean isCalendarBased() {
...
+        innerPeriod.getUnits().stream()
+            .map(
+                temporalUnit ->
+                    new Tuple<>(innerPeriod.get(temporalUnit), temporalUnit.getDuration()))
+            .mapToLong(
+                durationValuePair ->
+                    Math.multiplyExact(
+                        durationValuePair.getLeft(), durationValuePair.getRight().toMillis()))
+            .reduce(0, Math::addExact);
+    return Math.addExact(durationMillis, periodMillis);
```
Comment (korthout) [inline bpmn-model/src/main/java/io/camunda/zeebe/model/bpmn/util/time/Interval.java:198]: 👍 Nice work coming up with a solution here

I thought about this code for some time, and I understand its need. Periods and Durations don't combine well, and Java's standard time library doesn't allow easily converting the two to milliseconds. The problem is that Periods may be mapped to different Durations depending on the TimeZone. With this in mind, I worried that this implementation might be incorrect in some situations. But does it matter for the current use case? It is useful to tell the engine to backoff a job for several days, but this does not have to be precise. So, in my opinion, it's fine to keep it this way. Let me know how you feel about this.

🔧 In any case, let's document this, so the method is not used for cases where this precision is required for Intervals containing Periods.

File: bpmn-model/pom.xml [hunk 2]
```diff
@@ -33,6 +33,10 @@
       <groupId>org.camunda.bpm.model</groupId>
       <artifactId>camunda-xml-model</artifactId>
     </dependency>
+    <dependency>
+      <groupId>io.camunda</groupId>
+      <artifactId>zeebe-util</artifactId>
+    </dependency>
```
Comment (korthout) [inline bpmn-model/pom.xml:39]: ❌ We should not add this dependency. The Zeebe Util is not built for Java 8 and would push a lot of new transitive dependencies to consumers of the bpmn-model library. 

>**Note** it seems only to be needed for the Tuple class. Perhaps you can find a different solution without this class.

**Review by korthout — CHANGES_REQUESTED** [review]
Thanks for this awesome contribution @aivinog1 🚀 Great stuff!

>We could settle down the API and I could create a PR in the Modeler

👍 Both the BPMN API (XML) as well as the bpmn-model API (Java) look good to me.

❌ I found a few things that must be changed before this can be accepted.

Please have a look at my comments, suggestions and questions.

**Comment by nikku** [comment]
[Which direction](https://github.com/camunda/zeebe/issues/11101#issuecomment-1410297770) to you go with this PR?

* Back-off as an additional property?
* Back-off encoded in retries (be able to define a repeating interval), i.e. via ISO date 8601?

**Comment by aivinog1** [comment]
> [Which direction](https://github.com/camunda/zeebe/issues/11101#issuecomment-1410297770) to you go with this PR?
> 
> * Back-off as an additional property?
> * Back-off encoded in retries (be able to define a repeating interval), i.e. via ISO date 8601?

Hey @nikku!
Right now it is two separate attributes

**Comment by korthout** [comment]
> > [Which direction](https://github.com/camunda/zeebe/issues/11101#issuecomment-1410297770) to you go with this PR?
> > 
> > * Back-off as an additional property?
> > * Back-off encoded in retries (be able to define a repeating interval), i.e. via ISO date 8601?
> 
> Right now it is two separate attributes

For completeness, we made a decision about this in https://github.com/camunda/zeebe/issues/11101#issuecomment-1412328461:
> 📝 Let's decide to introduce the retriesBackoff as a secondary field next to retries.

**Comment by korthout** [comment]
Hi @aivinog1 👋 Just checking in with you to see if you're still interested in continuing this pull request. If you are, is there anything I can do to unblock you?

**Comment by aivinog1** [comment]
Hi @korthout 👋
Sure, I'm interested 🙂 But our team's priorities changed a bit and I'll come back to it in a couple of weeks 🙂

**Comment by korthout** [comment]
Thanks for the update @aivinog1. Looking forward to it, and let me know if I can assist in any way 🏗️

**Comment by npepinpe** [comment]
What's the status here? Lots of changes were done in the meantime to jobs.

**Comment by Zelldon** [comment]
This PR is open since a year. I will close it. Feel free to reopen or create another PR if you want to contribute this change.

**Pull request closed without merging**